nsq/reader.py

```diff
@@ -702,14 +709,31 @@ def process_message(self, message):
 
     def giving_up(self, message):
         """
-        Called when a message has been received where ``msg.attempts > max_tries``
+        Called when a message has been received where ``msg.attempts > max_tries``.
+        
+        Failed messages will be archived to ``$data_path/$channel_$topic/%Y%m%d-%H%M%S-%f_$sequence.failed.msg``
 
-        This is useful to subclass and override to perform a task (such as writing to disk, etc.)
+        This is useful to subclass and override to perform a custom task (such as writing to disk, etc.)
 
         :param message: the :class:`nsq.Message` received
         """
-        logger.warning('[%s] giving up on message %s after %d tries (max:%d) %r',
-                       self.name, message.id, message.attempts, self.max_tries, message.body)
+
+        self.failed_count += 1
+        
+        path = os.path.join(self.data_path or "", self.topic + '_' + self.channel)
+        if not os.path.exists(path):
+            os.makedirs(path)
+        
+        date_str = datetime.datetime.utcnow().strftime("%Y%m%d-%H%M%S-%f")
+        filename = "%s_%d.failed.msg" % (date_str, self.failed_count)
```

Comment:
i have mixed feelings. In some ways it could augment `self.failed_count` (which has some downsides across multiple).

however since `message.id` is source unique not globally unique (depends on nsqd workerID) putting in the filename might imply uniqueness that it shouldn't. Also, even though we create hex message id's, the protocol defines it as arbitrary bytes which wouldn't be safe for filename.

I chose to calculate the filename and include that in the log line as an alternative way a) raise awareness of the file archiving and b) link the two.